File: georocket-server/src/main/java/de/fhg/igd/georocket/input/XMLStartElement.java
package de.fhg.igd.georocket.input;

/**
 * A simple class describing an XML start element with optional prefix,
 * namespaces and attributes.
 * @author Michel Kraemer
 */
public class XMLStartElement {
  private final String prefix;
  private final String localName;
  private final String[] namespacePrefixes;
  private final String[] namespaceUris;
  private final String[] attributePrefixes;
  private final String[] attributeLocalNames;
  private final String[] attributeValues;
  
  /**
   * Create a new start element with a given name
   * @param localName the name
   */
  public XMLStartElement(String localName) {
    this(null, localName);
  }
  
  /**
   * Create a new start element with a given prefix and name
   * @param prefix the prefix (may be null)
   * @param localName the name
   */
  public XMLStartElement(String prefix, String localName) {
    this(prefix, localName, null, null);
  }
  
  /**
   * Create a new start element with a given prefix, name and namespaces
   * @param prefix the prefix (may be null)
   * @param localName the name
   * @param namespacePrefixes the namespace prefixes (may be null)
   * @param namespaceUris the namespace URIs (may be null)
   */
  public XMLStartElement(String prefix, String localName, String[] namespacePrefixes,
      String[] namespaceUris) {
    this(prefix, localName, namespacePrefixes, namespaceUris, null, null, null);
  }
  
  /**
   * Create a new start element with a given prefix, name and attributes
   * @param prefix the prefix (may be null)
   * @param localName the name
   * @param attributePrefixes the attribute prefixes (may be null)
   * @param attributeLocalNames the attribute names (may be null)
   * @param attributeValues the attribute values (may be null)
   */
  public XMLStartElement(String prefix, String localName, String[] attributePrefixes,
      String[] attributeLocalNames,  String[] attributeValues) {
    this(prefix, localName, null, null, attributePrefixes, attributeLocalNames, attributeValues);
  }
  
  /**
   * Create a new start element with a given prefix, name and namespaces
   * @param prefix the prefix (may be null)
   * @param localName the name
   * @param namespacePrefixes the namespace prefixes (may be null)
   * @param namespaceUris the namespace URIs (may be null)
   */
  public XMLStartElement(String prefix, String localName, String[] namespacePrefixes,
      String[] namespaceUris, String[] attributePrefixes, String[] attributeLocalNames,
      String[] attributeValues) {
    if ((namespacePrefixes == null || namespaceUris == null) && namespacePrefixes != namespaceUris) {
      throw new IllegalArgumentException("namespacePrefixes and namespaceUris must either be both null or non-null");
    }
    if ((attributePrefixes == null || attributeLocalNames == null || attributeValues == null) &&
        (attributePrefixes != attributeLocalNames || attributePrefixes != attributeValues)) {
      throw new IllegalArgumentException("attributePrefixes, attributeLocalNames and attributeValues "
          + "must either be all null or non-null");
    }
    if (namespacePrefixes != null && namespacePrefixes.length != namespaceUris.length) {
      throw new IllegalArgumentException("namespacePrefixes and namespaceUris must have the same number of elements");
    }
    if (attributePrefixes != null && (attributePrefixes.length != attributeLocalNames.length ||
        attributePrefixes.length != attributeValues.length)) {
      throw new IllegalArgumentException("attributePrefixes, attributeLocalNames and attributeValues "
          + "must have the same number of elements");
    }
    this.prefix = prefix;
    this.localName = localName;
    this.namespacePrefixes = namespacePrefixes;
    this.namespaceUris = namespaceUris;
    this.attributePrefixes = attributePrefixes;
    this.attributeLocalNames = attributeLocalNames;
    this.attributeValues = attributeValues;
  }
  
  /**
   * @return the element's prefix or null if it doesn't have one
   */
  public String getPrefix() {
    return prefix;
  }
  
  /**
   * @return the element's local name
   */
  public String getLocalName() {
    return localName;
  }
  
  /**
   * @return the elements name consisting of the prefix and the local name
   */
  public String getName() {
    if (prefix != null && !prefix.isEmpty()) {
      return prefix + ":" + localName;
    }
    return localName;
  }
  
  /**
   * @return the number of namespaces attached to this element
   */
  public int getNamespaceCount() {
    return namespacePrefixes != null ? namespacePrefixes.length : 0;
  }
  
  /**
   * Get a namespace prefix at a given position
   * @param i the position
   * @return the prefix
   */
  public String getNamespacePrefix(int i) {
    return namespacePrefixes[i];
  }
  
  /**
   * Get a namespace URI at a given position
   * @param i the position
   * @return the URI
   */
  public String getNamespaceUri(int i) {
    return namespaceUris[i];
  }
  
  /**
   * @return the number of attributes attached to this element
   */
  public int getAttributeCount() {
    return attributePrefixes != null ? attributePrefixes.length : 0;
  }
  
  /**
   * Get an attribute prefix at a given position
   * @param i the position
   * @return the prefix
   */
  public String getAttributePrefix(int i) {
    return attributePrefixes[i];
  }
  
  /**
   * Get the local name of an attribute at a given position
   * @param i the position
   * @return the local name
   */
  public String getAttributeLocalName(int i) {
    return attributeLocalNames[i];
  }
  
  /**
   * Get an attribute value at a given position
   * @param i the position
   * @return the value
   */
  public String getAttributeValue(int i) {
    return attributeValues[i];
  }
  
  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("<" + getName() + "");
    
    for (int i = 0; i < getNamespaceCount(); ++i) {
      sb.append(" xmlns");
      String prefix = getNamespacePrefix(i);
      if (prefix != null && !prefix.isEmpty()) {
        sb.append(":");
        sb.append(prefix);
      }
      sb.append("=\"");
      sb.append(getNamespaceUri(i));
      sb.append("\"");
    }
    
    for (int i = 0; i < getAttributeCount(); ++i) {
      sb.append(" ");
      String prefix = getAttributePrefix(i);
      if (prefix != null && !prefix.isEmpty()) {
        sb.append(prefix);
        sb.append(":");
      }
      sb.append(getAttributeLocalName(i));
      sb.append("=\"");
      sb.append(getAttributeValue(i));
      sb.append("\"");
    }
    
    sb.append(">");
    
    return sb.toString();
  }
}

File: kotlin_impl/src/main/kotlin/io/georocket/util/XMLStartElement.kt
package io.georocket.util

import com.fasterxml.jackson.annotation.JsonIgnore
import io.vertx.core.json.JsonObject
import io.vertx.kotlin.core.json.jsonObjectOf

/**
 * A simple class describing an XML start element with optional prefix,
 * namespaces and attributes.
 * @since 1.0.0
 * @author Michel Kraemer
 */
data class XMLStartElement constructor(
  val prefix: String? = null,
  val localName: String,
  val namespacePrefixes: List<String?> = emptyList(),
  val namespaceUris: List<String> = emptyList(),
  val attributePrefixes: List<String?> = emptyList(),
  val attributeLocalNames: List<String> = emptyList(),
  val attributeValues: List<String> = emptyList()
) {

  init {
    require(
      namespacePrefixes.size == namespaceUris.size
    ) {
      "namespacePrefixes and namespaceUris must have the same number of elements"
    }
    require(
      attributePrefixes.size == attributeLocalNames.size && attributePrefixes.size == attributeValues.size
    ) {
      "attributePrefixes, attributeLocalNames and attributeValues must have the same number of elements"
    }
    require(prefix == null  || prefix.isNotEmpty()) {
      "prefix must not be empty - to indicate that there is no prefix use the `null` value"
    }
    require(namespacePrefixes.all { it == null || it.isNotEmpty() }) {
      "namespace prefixes must not be empty - use the `null` value to refer to the default namespace"
    }
    require(attributePrefixes.all { it == null || it.isNotEmpty() }) {
      "attribute prefixes must not be empty - use the `null` value for attributes without prefix"
    }
  }

  /**
   * @return the elements name consisting of the prefix and the local name
   */
  @get:JsonIgnore
  val name: String
    get() = if (prefix != null) "$prefix:$localName" else localName

  /**
   * @return the number of namespaces attached to this element
   */
  @get:JsonIgnore
  val namespaceCount: Int
    get() = namespacePrefixes.size

  /**
   * Get a namespace prefix at a given position
   * @param i the position
   * @return the prefix
   */
  fun getNamespacePrefix(i: Int): String? {
    return namespacePrefixes[i]
  }

  /**
   * Get a namespace URI at a given position
   * @param i the position
   * @return the URI
   */
  fun getNamespaceUri(i: Int): String {
    return namespaceUris[i]
  }

  /**
   * @return the number of attributes attached to this element
   */
  @get:JsonIgnore
  val attributeCount: Int
    get() = attributePrefixes.size

  /**
   * Get an attribute prefix at a given position
   * @param i the position
   * @return the prefix
   */
  fun getAttributePrefix(i: Int): String? {
    return attributePrefixes[i]
  }

  /**
   * Get the local name of an attribute at a given position
   * @param i the position
   * @return the local name
   */
  fun getAttributeLocalName(i: Int): String {
    return attributeLocalNames[i]
  }

  /**
   * Get an attribute value at a given position
   * @param i the position
   * @return the value
   */
  fun getAttributeValue(i: Int): String {
    return attributeValues[i]
  }

  override fun toString(): String {
    val sb = StringBuilder()
    sb.append("<$name")
    for (i in 0 until namespaceCount) {
      sb.append(" xmlns")
      val prefix = getNamespacePrefix(i)
      if (prefix != null) {
        sb.append(":")
        sb.append(prefix)
      }
      sb.append("=\"")
      sb.append(getNamespaceUri(i))
      sb.append("\"")
    }
    for (i in 0 until attributeCount) {
      sb.append(" ")
      val prefix = getAttributePrefix(i)
      if (prefix != null) {
        sb.append(prefix)
        sb.append(":")
      }
      sb.append(getAttributeLocalName(i))
      sb.append("=\"")
      sb.append(getAttributeValue(i))
      sb.append("\"")
    }
    sb.append(">")
    return sb.toString()
  }

  /**
   * @return this object as a [JsonObject]
   */
  fun toJsonObject(): JsonObject {
    return jsonObjectOf(
      "prefix" to prefix,
      "localName" to localName,
      "namespacePrefixes" to namespacePrefixes,
      "namespaceUris" to namespaceUris,
      "attributePrefixes" to attributePrefixes,
      "attributeLocalNames" to attributeLocalNames,
      "attributeValues" to attributeValues,
    )
  }

  companion object {
    /**
     * Converts a [JsonObject] to a [XMLStartElement]
     * @param obj the [JsonObject] to convert
     * @return the [XMLStartElement]
     */
    fun fromJsonObject(obj: JsonObject): XMLStartElement {
      return XMLStartElement(
        prefix = obj.getString("prefix")?.takeIf { it.isNotEmpty() },
        localName = obj.getString("localName")!!,
        attributePrefixes = obj.getJsonArray("attributePrefixes").map { jsonPrefix -> jsonPrefix?.toString()?.takeIf { it.isNotEmpty() } },
        attributeLocalNames = obj.getJsonArray("attributeLocalNames").map(Any::toString),
        attributeValues = obj.getJsonArray("attributeValues").map(Any::toString),
        namespacePrefixes = obj.getJsonArray("namespacePrefixes").map { jsonPrefix -> jsonPrefix?.toString()?.takeIf { it.isNotEmpty() } },
        namespaceUris = obj.getJsonArray("namespaceUris").map(Any::toString)
      )

    }
  }
}
